### PR TITLE
Adds Daru::Vector#match and Daru::Vector#apply_where (issue #175)

### PR DESCRIPTION
**Pull request by athityakumar**
Rspec test added, has been written in the older format to main consistency. This will be easier to change while re-writing all tests to modern specifications file-wise. RDoc documentation has also been added. `Daru::Vector#match` uses Regexp match via the `=~` operator.

Tackles issue #175

### CODE REVIEW

**Comment by athityakumar** [comment]
@v0dro @zverok - Following the discussions on the issues #175 and #174 , `#match` support has been provided only to `Daru::Vector`. Do review the PR. :smile:

**Comment by athityakumar** [comment]
@v0dro @zverok @lokeshh - Following the discussions on the issues #175 and #174 , #match support has been provided only to Daru::Vector. Do review the PR. :smile:

**Comment by athityakumar** [comment]
@zverok @v0dro : Just wanted to remind you about this PR - kindly review whenever you're free. :slightly_smiling_face:

**Comment by zverok** [comment]
Reviewing it now.

File: lib/daru/vector.rb [hunk 1]
```diff
@@ -1141,6 +1159,20 @@ def jackknife(estimators, k=1) # rubocop:disable Metrics/AbcSize,Metrics/MethodL
...
+    #
+    # @param Regexp A regular matching expression. For example, +/weeks/+.
+    # @return [Array] Conatining either +nil+ or integer values, according to the match with the given +regexp+
+    # @example
+    #   dv = Daru::Vector.new ['3 days', '5 weeks', '2 weeks']
+    #   dv.match /weeks/
+    #
+    #   # => [nil, 2, 2]
+    def match(regexp)
+      @data.map { |x| x =~ regexp }
```
Comment (zverok) [inline lib/daru/vector.rb]: Let's use [===](http://www.rubyletter.com/blog/2017/01/18/ruby-triple-equals-operator.html) operator. This way we can also do `vector.match(1..20)` or `vector.match(Numeric)`.

File: spec/core/query_spec.rb [hunk 2]
```diff
@@ -334,4 +334,14 @@
...
   end
+  describe "apply_where" do
+    context Daru::Vector do
+      it "matches regexp with a block input" do
+        dv = Daru::Vector.new ['3 days', '5 weeks', '2 weeks']
+        dv = dv.apply_where(dv.match /weeks/) { |x| "#{x.split.first.to_i * 7} days" }
+        expected_dv = Daru::Vector.new ['3 days', '35 days', '14 days']
+        expect(dv).to eq expected_dv
+      end
+    end
```
Comment (zverok) [inline spec/core/query_spec.rb]: a) modern tests please (with `subject` and one-statement checks, see betterspecs.org)
b) you need to check several conditions, like "empty boolean vector", "block returning something unexpected" and so on.

File: spec/vector_spec.rb [hunk 3]
```diff
@@ -2009,6 +2009,13 @@ def expect_correct_tokens hash
     end
   end
 
+  context '#match' do
+      it 'returns matching array for a given regexp' do
+        dv = Daru::Vector.new ['3 days', '5 weeks', '2 weeks']
+        expect(dv.match /weeks/).to eq [nil, 2, 2]
+      end
```
Comment (zverok) [inline spec/vector_spec.rb]: Same as above: modern specs and several conditions.

Comment (zverok) [inline spec/vector_spec.rb]: BTW, if I understand correctly, it is better to have only `false` and `true` values returned (though, if you'll follow `===` advice, you will have this automatically).

File: lib/daru/core/query.rb [hunk 4]
```diff
@@ -70,6 +70,23 @@ def vector_where dv, bool_array
           resultant_dv
         end
 
+        def vector_apply_where dv, bool_array, &block
+          _data, new_index = fetch_new_data_and_index dv, bool_array
+          all_index = dv.index.to_a
+          yield_data = dv.map(&block).to_a
+          all_data = all_index.map { |idx| new_index.include?(idx) ? yield_data[idx] : dv[idx] }
```
Comment (zverok) [inline lib/daru/core/query.rb]: Are you sure it is a sane approach (apply block to all data, and then select appropriate)? 
Imagine there is 10k values vector, and "where" returns only 5 values of all 10k, and calculation in `block` is expensive (some serious math, or even reading some files). "Right" implementation should call block 5 times, not 10k times and then select only 5 values.

**Comment by v0dro** [comment]
Whats the update with the PR?

**Comment by v0dro** [comment]
Please finish it off ASAP.

File: spec/core/query_spec.rb [hunk 5]
```diff
@@ -334,4 +334,14 @@
       end
     end
   end
+
+  describe "apply_where" do
+    context "matches regexp with block input" do
+      subject { dv.apply_where(dv.match /weeks/) { |x| "#{x.split.first.to_i * 7} days" } }
```
Comment (athityakumar) [inline spec/core/query_spec.rb]: @zverok - I remember seeing DSL support being added to `which` in PR #396. Should a similar DSL call be supported here, like `apply_which`?

**Comment by athityakumar** [comment]
@zverok - Please review when you're free. 😄

**Comment by v0dro** [comment]
I think @zverok is busy with refactoring :P

**Pull request merged**